### PR TITLE
Add an EVP demo for CMAC

### PR DESCRIPTION
**Pull request by faramir-dev**
Fixes openssl#14110

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Review by paulidale — APPROVED** [review]

**Review by t8m — APPROVED** [review]

File: demos/mac/cmac-aes256.c [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    size_t out_len = 0;
+    OSSL_PARAM params[4], *p = params;
+    char cipher_name[] = "aes256";
+
+    library_context = OSSL_LIB_CTX_new();
+    if (library_context == NULL) {
+        fprintf(stderr, "OSSL_LIB_CTX_new() returned NULL\n");
+        goto end;
+    }
+    OSSL_LIB_CTX_set0_default(library_context);
```
Comment (slontis) [inline demos/mac/cmac-aes256.c:82]: I would prefer that the OSSL_LIB_CTX_set0_default is not used in demos.. It is really a hacky function for swapping the libctx temporarily for API's that do not correctly handle the libctx..

File: demos/mac/Makefile [hunk 2]
```diff
@@ -15,10 +15,11 @@ all: gmac poly1305
```
Comment (slontis) [inline demos/mac/Makefile:15]: The all: line should include cmac and hmac

File: demos/mac/cmac-aes256.c [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+    rv = EXIT_SUCCESS;
+end:
+    if (rv != EXIT_SUCCESS)
+        ERR_print_errors_fp(stderr);
+    /* OpenSSL free functions will ignore NULL arguments */
+    OPENSSL_free(out);
+    EVP_MAC_CTX_free(mctx);
+    EVP_MAC_free(mac);
+    OSSL_LIB_CTX_free(library_context);
```
Comment (slontis) [inline demos/mac/cmac-aes256.c:153]: Use valgrind.. This leaks currently. OSSL_LIB_CTX_set0_default() is probably the reason.

File: demos/mac/cmac-aes256.c [hunk 4]
```diff
@@ -66,23 +66,15 @@ static const char *propq = NULL;
...
     size_t out_len = 0;
     OSSL_PARAM params[4], *p = params;
     char cipher_name[] = "aes256";
 
-    library_context = OSSL_LIB_CTX_new();
-    if (library_context == NULL) {
-        fprintf(stderr, "OSSL_LIB_CTX_new() returned NULL\n");
-        goto end;
-    }
-    OSSL_LIB_CTX_set0_default(library_context);
```
Comment (paulidale) [inline demos/mac/cmac-aes256.c]: I think Shane meant to just omit this line.

Comment (faramir-dev) [reply demos/mac/cmac-aes256.c]: `EVP_MAC_init` is not able to find `AES256` without this line. :-(

Comment (paulidale) [reply demos/mac/cmac-aes256.c]: I'd call that a bug in CMAC ...

Comment (paulidale) [reply demos/mac/cmac-aes256.c]: It is found here when I try.

Comment (paulidale) [reply demos/mac/cmac-aes256.c]: It also works when I force the null provider into the default library context & use a separate lib ctx for the MAC operation.

**Review by t8m — APPROVED** [review]

**Review by slontis — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master and 3.0 branches. Thank you.

**Pull request closed without merging**